metadata/src/main/java/org/apache/kafka/controller/LocalLogManager.java

```diff
@@ -0,0 +1,798 @@
+ *                            V                V
+ *                 +-------------------------------------------+
+ *                 |         LeadershipClaimerThread           |
+ *                 +-------------------------------------------+
+ *                      | blocking | blocking |
+ *                      | claim    | renounce | beginShutdown
+ *                      V          V          V
+ *                +-------------------------------------------+      +----------+
+ * maybeWrite --} |               ScribeThread                | ---} | Listener |
+ *                +-------------------------------------------+      +----------+
+ *                      |        |          ^       |
+ *                      | pause  | unpause  | wake  | beginShutdown
+ *                      V        V          |       V
+ *                +-------------------------------------------+
+ *                |               WatcherThread               |
+ *                +-------------------------------------------+
+ */
+public final class LocalLogManager implements MetaLogManager {
+    /**
+     * The global registry of locks on a particular path.
+     */
+    private static final LockRegistry LOCK_REGISTRY = new LockRegistry();
+
+    /**
+     * An empty byte buffer object.
+     */
+    private static final ByteBuffer EMPTY = ByteBuffer.allocate(0);
+
+    /**
+     * A lock plus the number of threads waiting on it.  Waiters must be updated while
+     * holding the lock of the LockRegistry itself.
+     */
+    static class LockData {
+        final ReentrantLock lock;
+        int waiters;
+
+        LockData() {
+            this.lock = new ReentrantLock();
+            this.waiters = 0;
+        }
+    }
+
+    /**
+     * A registry for locks on paths.
+     *
+     * We want a lock that works in a cross-process fashion, so that multiple processes
+     * on the same machine can attempt to become the leader of our file-backed mock log.
+     *
+     * FileChannel#lock is almost good enough to do the job on its own, but it has a
+     * quirk: if two threads from the same process try to lock the same file, one of them
+     * will get an OverlappingFileLockException.  This would be bad for the case where
+     * we run multiple controllers in the same process in JUnit.  So we add another layer
+     * of in-memory locking to ensure that we don't hit this case.
+     */
+    static class LockRegistry {
+        private final Map<String, LockData> locks = new HashMap<>();
+
+        void lock(String path, FileChannel channel, Runnable onTakeLock)
+                throws IOException {
+            try {
+                LockData lockData = null;
+                FileLock fileLock = null;
+                synchronized (this) {
+                    lockData = locks.computeIfAbsent(path, __ -> new LockData());
+                    lockData.waiters++;
+                }
+                lockData.lock.lockInterruptibly();
```

Comment:
I think this could throw `InterruptedException` in which case `waiters--` would still not get called?
```suggestion
                try {
                    lockData.lock.lockInterruptibly();
                } catch (InterruptedException e) {
                    lockData.waiters--;
                    throw e;
                }
```

Reply:
It would need to be more than that (`synchronized(this)`), but I assume you know what I'm getting at.

Reply:
fair point.  I fixed this again, hopefully this time for good.